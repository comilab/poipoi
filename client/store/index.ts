import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
