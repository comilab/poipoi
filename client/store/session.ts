import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
