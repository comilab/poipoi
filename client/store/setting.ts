import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
