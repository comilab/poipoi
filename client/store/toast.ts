import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
