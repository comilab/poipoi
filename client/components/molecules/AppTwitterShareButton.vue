<template lang="pug">
AppButton.tweet-button(
  :href="twitterShareUrl",
  color="blue",
  :icon="faTwitter",
  target="_blank",
  v-bind="$attrs"
) Tweet
</template>

<script lang="ts">
import { defineComponent, computed } from 'nuxt-composition-api'
import { faTwitter } from '@fortawesome/free-brands-svg-icons'

import { useStore } from '~/store'
import AppButton from '~/components/atoms/AppButton.vue'

export default defineComponent({
  components: {
    AppButton
  },
  props: {
    url: {
      type: String,
      required: false,
      default: ''
    },
    text: {
      type: String,
      required: false,
      default: ''
    },
    appendSiteTitle: {
      type: Boolean,
      required: false,
      default: true
    }
  },
  setup (props) {
    const store = useStore()

    const twitterShareUrl = computed(() => {
      const url = new URL('http://twitter.com/share')
      const texts = []

      let maxLength = 140
      if (props.url) {
        maxLength -= 11
        url.searchParams.append('url', props.url)
      }

      if (props.appendSiteTitle) {
        const siteTitle = store.setting.data.value!.siteTitle
        maxLength -= siteTitle.length
        if (props.text.length) {
          maxLength -= 3
          texts.push(' - ')
        }
        texts.push(siteTitle)
      }

      if (props.text.length > maxLength) {
        maxLength -= 1
        texts.unshift(`${props.text.substring(0, maxLength)}â€¦`)
      } else {
        texts.unshift(props.text.substring(0, maxLength))
      }

      url.searchParams.append('text', texts.join(''))
      return url.toString()
    })

    return {
      twitterShareUrl,
      faTwitter
    }
  }
})
</script>

<style lang="postcss" scoped>
.tweet-button {
  background-color: #1da1f2;

  &:hover {
    @apply bg-blue-600;
  }
}
</style>
