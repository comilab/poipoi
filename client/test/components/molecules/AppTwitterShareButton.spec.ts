import { render } from '@testing-library/vue'
import { defineComponent } from 'nuxt-composition-api'
import _ from 'lodash'
import { faTwitter } from '@fortawesome/free-brands-svg-icons'

import AppTwitterShareButton from '~/components/molecules/AppTwitterShareButton.vue'
import { StoreProvider, store } from '~/store'
import settingFactory from '~/test/factories/setting'

describe('components/molecules/AppTwitterShareButton', () => {
  let props: any
  let wrapperComponent: any
  let options: any

  beforeEach(() => {
    store.setting.setData(settingFactory.build({
      siteTitle: 'site-title'
    }))
    props = {
      appendSiteTitle: false
    }
    wrapperComponent = defineComponent({
      components: { StoreProvider, AppTwitterShareButton },
      setup () {
        return { attrs: props }
      },
      template: `
        <StoreProvider>
          <AppTwitterShareButton v-bind="attrs" />
        </StoreProvider>
      `
    })
    options = {
      props: {},
      stubs: {
        FontAwesomeIcon: true
      }
    }
  })

  test('renders correctly', () => {
    const { container } = render(wrapperComponent, options)

    expect(container).toMatchSnapshot()
  })

  describe('props', () => {
    let buttonProps: any

    beforeEach(() => {
      buttonProps = {}
      options.stubs.AppButton = defineComponent({
        props: ['href'],
        setup (props) {
          buttonProps = props
        },
        template: '<div />'
      })
    })

    describe('url', () => {
      test('urlクエリを追加', () => {
        props.url = 'https://example.com'
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('url')).toBe(props.url)
      })
    })

    describe('text', () => {
      test('textパラメータを追加', () => {
        props.text = 'foo'
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')).toBe(props.text)
      })

      test('140文字に丸める', () => {
        props.text = _.times(141, String).join('')
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')!.length).toBe(140)
        expect(url.searchParams.get('text')!.substr(-1)).toBe('…')
      })

      test('urlがある場合は140-11文字に丸める', () => {
        props.text = _.times(141, String).join('')
        props.url = 'https://example.com'
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')!.length).toBe(140 - 11)
      })
    })

    describe('appendSiteTitle', () => {
      beforeEach(() => {
        store.setting.setData(settingFactory.build({
          siteTitle: 'site-title'
        }))
      })

      test('trueの場合は" - {サイト名}"をtext末尾に付与', () => {
        props.text = 'text'
        props.appendSiteTitle = true
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')).toBe('text - site-title')
      })

      test('trueかつtextが空の場合はサイト名のみをtextにセット', () => {
        props.appendSiteTitle = true
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')).toBe('site-title')
      })

      test('text全体で140文字に丸まるようにする', () => {
        props.text = _.times(141, String).join('')
        props.appendSiteTitle = true
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')!.length).toBe(140)
      })

      test('falseの場合はサイト名を付与しない', () => {
        props.text = 'text'
        props.appendSiteTitle = false
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')).toBe('text')
      })

      test('デフォルトはtrue', () => {
        props = {}
        render(wrapperComponent, options)

        const url = new URL(buttonProps.href)
        expect(url.searchParams.get('text')).toBe('site-title')
      })
    })
  })

  describe('AppButton', () => {
    let buttonProps: any

    beforeEach(() => {
      buttonProps = {}
      options.stubs.AppButton = defineComponent({
        props: ['href', 'color', 'icon', 'target'],
        setup (props) {
          buttonProps = props
        },
        template: '<div><slot /></div>'
      })
    })

    test('hrefにurlをセット', () => {
      render(wrapperComponent, options)

      expect(buttonProps.href).toMatch('http://twitter.com/share')
    })

    test('color = blue', () => {
      render(wrapperComponent, options)

      expect(buttonProps.color).toBe('blue')
    })

    test('icon = faTwitter', () => {
      render(wrapperComponent, options)

      expect(buttonProps.icon).toStrictEqual(faTwitter)
    })

    test('target = _blank', () => {
      render(wrapperComponent, options)

      expect(buttonProps.target).toBe('_blank')
    })
  })
})
