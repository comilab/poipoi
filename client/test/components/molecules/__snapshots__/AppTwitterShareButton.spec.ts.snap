// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppTwitterShareButton renders correctly 1`] = `
<div>
  <div>
    <a
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none tweet-button bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 rounded"
      data-testid="button"
      href="http://twitter.com/share?text="
      target="_blank"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw mr-2"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      >
        Tweet
      </div>
    </a>
  </div>
</div>
`;
